analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/PubliclyWritableDirectoriesBase.cs

```diff
@@ -0,0 +1,132 @@
+        private static readonly string DevShmRegex = string.Format(TwoPartPatternWithVariableSlash, "dev", "shm");
+        private static readonly string DevMqueueRegex = string.Format(TwoPartPatternWithForwardSlash, "dev", "mqueue");
+        private static readonly string RunLockRegex = string.Format(TwoPartPatternWithForwardSlash, "run", "lock");
+        private static readonly string VarRunLockRegex = string.Format(ThreePartPatternWithForwardSlash, "var", "run", "lock");
+        private static readonly string LibraryCachesRegex = string.Format(TwoPartPatternWithForwardSlash, "library", "caches");
+        private static readonly string UsersSharedRegex = string.Format(TwoPartPatternWithForwardSlash, "users", "shared");
+        private static readonly string PrivateTmpRegex = string.Format(TwoPartPatternWithForwardSlash, "private", "tmp");
+        private static readonly string PrivateVarTmpRegex = string.Format(ThreePartPatternWithForwardSlash, "private", "var", "tmp");
+        private static readonly string TempEnvVariable = string.Format(EnvironmentVariableTemplate, "temp");
+        private static readonly string TmpEnvVariable = string.Format(EnvironmentVariableTemplate, "tmp");
+        private static readonly string TmpDirEnvVariable = string.Format(EnvironmentVariableTemplate, "tmpdir");
+
+        private static readonly Regex LinuxPubliclyWritabelDirs =
+            new Regex($"{DevMqueueRegex}|{RunLockRegex}|{VarRunLockRegex}",
+                RegexOptions.Compiled);
+
+        private static readonly Regex WindowsAndMacPubliclyWritabelDirs =
+            new Regex($"{TmpRegex}|{VarTmpRegex}|{UsrTmpRegex}|{DevShmRegex}|{LibraryCachesRegex}|{UsersSharedRegex}|{PrivateTmpRegex}|{PrivateVarTmpRegex}|{TempRegex}|{WindowsTempRegex}",
+                RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
+
+        private static readonly Regex EnvironmentVariables =
+            new Regex($"{UserProfileRegex}|{TempEnvVariable}|{TmpEnvVariable}|{TmpDirEnvVariable}",
+                RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
+
+        protected abstract ILanguageFacade<TSyntaxKind> Language { get; }
+
+        private protected abstract bool IsGetTempPathAssignment(TInvocationExpression invocationExpression, KnownType type, string methodName, SemanticModel semanticModel);
+        private protected abstract bool IsInsecureEnvironmentVariableRetrieval(TInvocationExpression invocation, KnownType type, string methodName, SemanticModel semanticModel);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        protected DiagnosticDescriptor Rule { get; }
+
+        protected static readonly string[] InsecureEnvironmentVariables =
+            new[]
+                {
+                    "tmp",
+                    "temp",
+                    "tmpdir"
+                };
```

Comment:
Can be simplified a bit: `protected static readonly string[] InsecureEnvironmentVariables = {"tmp", "temp", "tmpdir"};`